src/read.c

```diff
@@ -1240,6 +1240,151 @@ void ReadRecExpr (
+**
+**  'ReadFuncArgList' reads the argument list of a function. In case of an error
+**  it skips all symbols up to one contained in <follow>.
+**
+**  <ArgList> :=    ('readwrite'|'readonly') <Ident>
+**                   {',' ('readwrite'|'readonly') <Ident> } ( '...' )
+**
+**  is_atomic: Is this an atomic function?
+**  symbol: The end symbol of the arglist (usually S_RBRACK, but S_RBRACE for
+**          lambda functions.
+**  symbolstr: symbol as an ascii string
+**
+**  This function assumes the opening bracket is already read, and is
+**  responsible for reading the closing bracket.
+*/
+
+struct ArgList ReadFuncArgList(
+    TypSymbolSet        follow,
+    Int is_atomic,
+    UInt symbol,
+    const Char * symbolstr)
+{
+    Obj        name;           /* one local variable name         */
+    Int        narg;           /* number of arguments             */
+    int        lockmode;       /* type of lock for current argument */
+    Obj        nams;           /* list of local variables names   */
+    Bag        locks = 0;      /* locks of the function */
+    UInt       isvarg = 0;     /* does function have varargs?     */
+
+    UInt       i;              /* loop variable */
+    if (is_atomic)
+        locks = NEW_STRING(4);
+
+    /* make and push the new local variables list (args and locals)        */
+    narg = 0;
+    nams = NEW_PLIST( T_PLIST, narg );
+    SET_LEN_PLIST( nams, narg );
+    TLS(CountNams) += 1;
+    ASS_LIST( TLS(StackNams), TLS(CountNams), nams );
+    if ( TLS(Symbol) != symbol ) {
+        lockmode = 0;
+        switch (TLS(Symbol)) {
+            case S_READWRITE:
+            if (!is_atomic) {
+                SyntaxError("'readwrite' argument of non-atomic function");
+                GetSymbol();
+                break;
+            }
+            lockmode++;
+            case S_READONLY:
+            if (!is_atomic) {
+                SyntaxError("'readonly' argument of non-atomic function");
+                GetSymbol();
+                break;
+            }
+            lockmode++;
+            CHARS_STRING(locks)[0] = lockmode;
+            SET_LEN_STRING(locks, 1);
+            GetSymbol();
+        }
+        C_NEW_STRING_DYN( name, TLS(Value) );
+        narg += 1;
+        ASS_LIST( nams, narg, name );
+        Match(S_IDENT,"identifier",symbol|S_LOCAL|STATBEGIN|S_END|follow);
+    }
+
+    if(TLS(Symbol) == S_DOTDOT) {
+        SyntaxError("Three dots required for variadic argument list");
+    }
+    if(TLS(Symbol) == S_DOTDOTDOT) {
+        isvarg = 1;
+        GetSymbol();
+    }
+
+    while ( TLS(Symbol) == S_COMMA ) {
+        if (isvarg) {
+            SyntaxError("Only final argument can be variadic");
+        }
+
+        Match( S_COMMA, ",", follow );
+        lockmode = 0;
+        switch (TLS(Symbol)) {
+            case S_READWRITE:
+            if (!is_atomic) {
+                SyntaxError("'readwrite' argument of non-atomic function");
+                GetSymbol();
```

Comment:
I added tests for this, and also for lambda (at the moment they are never valid for lambda, as you can't put the initial `atomic`. We could invent a notation for that, if we ever want one).